hbase-client/src/main/java/org/apache/hadoop/hbase/client/RawAsyncHBaseAdmin.java

```diff
@@ -3521,20 +3522,33 @@ private void getProcedureResult(long procId, CompletableFuture<Void> future, int
       return failedFuture(e);
     }
     CompletableFuture<Void> future = new CompletableFuture<>();
-    addListener(ConnectionFactory.createAsyncConnection(peerConf), (conn, err) -> {
+    CompletableFuture<AsyncConnection> cf = ConnectionFactory.createAsyncConnection(peerConf);
+    addListener(cf, (conn, err) -> {
       if (err != null) {
-        future.completeExceptionally(err);
+        try {
+          future.completeExceptionally(err);
+        } finally {
+          IOUtils.closeQuietly(conn);
```

Comment:
This method has been deprecated IIRC? Can use Closeables.close instead.

And maybe we could just call addListener on the above 'future', to close the AsyncConnection? Then we do not need to add close code everwhere.

Reply:
+1 for the above suggestion.

Reply:
Something similar to this might help avoid closing conn at each step:

```
.............
    CompletableFuture<Void> future = new CompletableFuture<>();
    CompletableFuture<AsyncConnection> asyncConnection =
      ConnectionFactory.createAsyncConnection(peerConf);
    addListener(asyncConnection, (conn, err) -> {
      future.whenComplete((resp, error) -> {
        try {
          Closeables.close(conn, true);
        } catch (IOException e) {
          throw new AssertionError("IOException should not be thrown for swallowIOException=true");
        }
      });
      if (err != null) {
        future.completeExceptionally(err);
        return;
      }
.............
```